waiter/src/waiter/scheduler/composite.clj

```diff
@@ -0,0 +1,165 @@
+
+  (create-service-if-new [_ descriptor]
+    (-> descriptor
+        :service-id
+        service-id->scheduler
+        (scheduler/create-service-if-new descriptor)))
+
+  (delete-service [_ service-id]
+    (-> service-id
+        service-id->scheduler
+        (scheduler/delete-service service-id)))
+
+  (scale-service [_ service-id target-instances force]
+    (-> service-id
+        service-id->scheduler
+        (scheduler/scale-service service-id target-instances force)))
+
+  (retrieve-directory-content [_ service-id instance-id host directory]
+    (-> service-id
+        service-id->scheduler
+        (scheduler/retrieve-directory-content service-id instance-id host directory)))
+
+  (service-id->state [_ service-id]
+    (-> service-id
+        service-id->scheduler
+        (scheduler/service-id->state service-id)))
+
+  (state [_]
+    (pc/map-vals scheduler/state scheduler-id->scheduler)))
+
+(defn service-id->scheduler
+  "Resolves the scheduler for a given service-id using the scheduler defined in the description.
+   If the scheduler is not specified, the default scheduler is returned."
+  [service-id->service-description-fn scheduler-id->scheduler default-scheduler-id service-id]
+  (let [service-description (service-id->service-description-fn service-id)
+        scheduler-id (get service-description "scheduler" default-scheduler-id)]
+    (if-let [scheduler (scheduler-id->scheduler scheduler-id)]
+      scheduler
+      (throw (ex-info "No matching scheduler found!"
+                      {:available-schedulers (-> scheduler-id->scheduler keys sort)
+                       :service-id service-id
+                       :specified-scheduler scheduler-id})))))
+
+(defn invoke-component-factory
+  "Creates a component based on the factory-fn specified in the component-config."
+  [context {:keys [factory-fn] :as component-config}]
+  (if-let [resolved-fn (utils/resolve-symbol factory-fn)]
+    (resolved-fn (merge context component-config))
+    (throw (ex-info "Unable to resolve factory function" (assoc component-config :ns (namespace factory-fn))))))
+
+(def component-schema
+  {s/Keyword {(s/required-key :factory-fn) s/Symbol
+              s/Keyword s/Any}})
+
+(defn initialize-component-schedulers
+  "Initializes individual component schedulers."
+  [{:keys [components] :as config}]
+  {:pre [(seq components)]}
+  (s/validate component-schema components)
+  (let [context (dissoc config :components)]
+    (->> components
+         (pc/map-vals #(invoke-component-factory context %))
+         (pc/map-keys name))))
+
+(defn create-composite-scheduler
+  "Creates and starts composite scheduler with components using their respective factory functions."
+  [{:keys [service-description-defaults service-id->service-description-fn] :as config}]
+  (let [scheduler-id->scheduler (initialize-component-schedulers config)
+        default-scheduler-id (get service-description-defaults "scheduler")
```

Comment:
Defaults shouldn't be read directly but instead read through the service-description.  It is elsewhere's responsibility to merge in the default if not specified by the user.

Reply:
What if we wanted the "default" scheduler to be based on something else in the `service-description`?